modules/standard/Memory/Initialization.chpl

```diff
@@ -0,0 +1,290 @@
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  The :mod:`Initialization` module provides functions which enable users to
+  move-initialize and deinitialize values.
+
+  The functionality provided by this module can be used to implement
+  collections in a manner similar to those implemented by the Chapel standard
+  modules (such as :mod:`List` or :mod:`Map`).
+*/
+module Initialization {
+
+  private inline proc _move(ref dst: ?t, const ref src: t) {
+    __primitive("=", dst, src);
+  }
+
+  private inline proc _destroy(ref r) {
+    chpl__autoDestroy(r);
+  }
+
+  private inline proc _needsAutoDestroy(type t) param: bool {
+    return __primitive("needs auto destroy", t);
+  }
+
+  /*
+    Check to see if a given type supports deinitialization.
+
+    :arg t: A type to check for a deinitializer
+    :type t: `type`
+
+    :return: ``true`` if ``t`` supports deinitialization
+    :rtype: param bool
+  */
+  proc needsDeinit(type t) param {
+    return _needsAutoDestroy(t);
+  }
+
+  /*
+    Explicitly deinitialize a reference. The memory referred to by ``arg``
+    should be considered uninitialized after a call to this function.
+
+    .. warning::
+
+      Chapel is not aware of a call to ``explicitDeinit()``, which means
+      that the deinitializer of ``arg`` may run again once it goes out of
+      scope.
+
+    :arg: A value to deinitialize
+  */
+  proc explicitDeinit(ref arg: ?t) {
+    if needsDeinit(t) then
+      _destroy(arg);
+  }
+
+  /*
+    Move-initialize a reference from a value. No assignment is performed
+    when moving ``rhs`` into the memory referred to by ``lhs``.
+
+    .. warning::
+
+      If the value at ``lhs`` is already initialized memory, it will be
+      overwritten by the contents of ``rhs`` without being deinitialized.
+      Call :proc:`explicitDeinit()` to deinitialize ``lhs`` first if
+      necessary.
+
+    :arg lhs: A reference to move-initialize
+    :type lhs: `?t`
+
+    :arg rhs: A 
+    :type rhs: `t`
+  */
+  proc moveInitialize(ref lhs: ?t, pragma "no auto destroy" in rhs: t) {
+    if lhs.type != nothing then
+      _move(lhs, rhs);
+  }
+
+  /*
+    Move the memory referred to by a reference into a new value.
+
+    .. warning::
+
+      The memory referred to by ``arg`` should be considered uninitialized
+      after a call to this function.
+
+    :arg arg: A reference to consume
+    :type arg: `?t`
+
+    :return: A reference moved into a new value
+    :rtype: `t`
+  */
```

Comment:
I think this should be called `moveToReturn`.
Things can be moved into "variables", not into "values".
I can move something into an integer, for example,
only because here "integer" is a slang for an "integer variable". :)

Not sure how to write the chpldoc... perhaps:

    Give ownership of the referenced value to the caller.

We cannot say "pass" the ownership, as it is not really taken away from the argument.

:arg arg: the value to consume [the reader will see that it is passed by const-ref]
:return: the value referenced by the argument

Reply:
I think the line between variable and value here is kind of thin, because what if I do something like...

```chapel
// What about this?
moveToValue(foobar);

// As opposed to this?
var x = moveToValue(foobar);
```

Here the function `moveToValue` is returning a variable moved into a new value, which may or may not be assigned to a `var`.

The name `moveToReturn` doesn't make much sense to me because it describes what the function is doing, not what it is used for. That we move some memory into a temporary variable within the function and then return it does not matter, because once returned it may or may not be captured into a variable.

Reply:
We should probably move further discussion to #16172 though.